### ISSUE TITLE
False positive in `clippy::panicking_unwrap`

### ISSUE THREAD

**Issue by pacak** [rust-lang/rust-clippy#16436]
### Summary

Clippy absolutely rejects some strange looking but not panicking code.

### Lint Name

clippy::panicking_unwrap

### Reproducer

I tried this code:

```rust
struct Foo {
    inner: Option<bool>,
}

impl Foo {
    fn poke(&mut self) -> bool {
        if self.inner.is_some() {
            self.inner.unwrap()
        } else {
            self.populate();
            self.inner.unwrap()
        }
    }

    fn populate(&mut self) {
        self.inner = Some(false);
    }
}
```

I saw this happen:

```
error: this call to `unwrap()` will always panic
  --> src/lib.rs:11:13
   |
 7 |         if self.inner.is_some() {
   |            -------------------- because of this check
...
11 |             self.inner.unwrap()
   |             ^^^^^^^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.93.0/index.html#panicking_unwrap
   = note: `#[deny(clippy::panicking_unwrap)]` on by default
```

I expected to see this happen:

This code won't panic so at least it shouldn't be denied.

### Version

```text

```

**Comment by pacak** [comment]
^ That's for pre-release testing of 1.93

**Comment by theemathas** [comment]
This is probably the same bug as #16188. It's fixed in nightly, so it will be fixed in stable in 1.94.0, unless a point release happens. It's probably currently too late to get this into 1.93.0.

**Comment by flip1995** [comment]
_If_ there is a point release 1.93.1 we'll include the fix for this there. If not, this fix will be released with 1.94.0. Sorry for the inconvenience.
